In [1]:
from Bio import SeqIO, Seq
import tqdm

# OLD 2024 MODEL (YF_IDs)

In [2]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_cds.fasta', 'r') as f:
    cds_records = list(SeqIO.parse(f, 'fasta'))

cds = [r.id for r in cds_records]
cds[0], len(cds), len(set(cds))

('YF00013476.t1', 25987, 25987)

In [3]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_exon.fasta', 'r') as f:
    exon_records = list(SeqIO.parse(f, 'fasta'))

exon = [r.id for r in exon_records]
exon[0]
exon[0], len(exon), len(set(exon))

('YF00013476.t1', 26749, 26749)

### THE DIFFERENCE BETWEEN THE EXON AND CDS FILES

In [4]:
len(set.difference(set(exon), set(cds))), len(set.difference(set(cds), set(exon)))

(1104, 342)

1104 more genes in the exon file.
342 more genes in the cds file.

In [5]:
set.difference(set(exon), set(cds))

{'YF00036599.t1',
 'YF00036135.t1',
 'YF00037746.t1',
 'YF00035153.t1',
 'YF00034129.t1',
 'YF00037559.t1',
 'YF00034350.t1',
 'YF00036011.t1',
 'YF00036936.t1',
 'YF00035395.t1',
 'YF00036068.t1',
 'YF00034603.t1',
 'YF00034061.t1',
 'YF00036334.t1',
 'YF00037123.t1',
 'YF00036131.t1',
 'YF00037500.t1',
 'YF00036866.t1',
 'YF00038353.t1',
 'YF00037780.t1',
 'YF00037352.t1',
 'YF00035408.t1',
 'YF00034545.t1',
 'YF00035092.t1',
 'YF00034074.t1',
 'YF00038199.t1',
 'YF00037373.t1',
 'YF00034853.t1',
 'YF00037984.t1',
 'YF00037547.t1',
 'YF00037393.t1',
 'YF00037925.t1',
 'YF00037080.t1',
 'YF00037061.t1',
 'YF00036848.t1',
 'YF00036076.t1',
 'YF00035967.t1',
 'YF00034914.t1',
 'YF00035455.t1',
 'YF00036044.t1',
 'YF00037878.t1',
 'YF00038049.t1',
 'YF00036209.t1',
 'YF00035490.t1',
 'YF00037157.t1',
 'YF00037895.t1',
 'YF00036437.t1',
 'YF00037531.t1',
 'YF00035449.t1',
 'YF00036314.t1',
 'YF00037442.t1',
 'YF00037595.t1',
 'YF00035954.t1',
 'YF00035317.t1',
 'YF00035676.t1',
 'YF000370

In [6]:
set.difference(set(cds), set(exon))

{'YF00034729.t1',
 'YF00034884.t1',
 'YF00035084.t1',
 'YF00035086.t1',
 'YF00035093.t1',
 'YF00035114.t1',
 'YF00035132.t1',
 'YF00035136.t1',
 'YF00035163.t1',
 'YF00035177.t1',
 'YF00035203.t1',
 'YF00035243.t1',
 'YF00035247.t1',
 'YF00035249.t1',
 'YF00035279.t1',
 'YF00035288.t1',
 'YF00035307.t1',
 'YF00035318.t1',
 'YF00035331.t1',
 'YF00035336.t1',
 'YF00035349.t1',
 'YF00035368.t1',
 'YF00035376.t1',
 'YF00035390.t1',
 'YF00035399.t1',
 'YF00035407.t1',
 'YF00035409.t1',
 'YF00035417.t1',
 'YF00035420.t1',
 'YF00035431.t1',
 'YF00035456.t1',
 'YF00035461.t1',
 'YF00035464.t1',
 'YF00035468.t1',
 'YF00035488.t1',
 'YF00035497.t1',
 'YF00035503.t1',
 'YF00035512.t1',
 'YF00035527.t1',
 'YF00035536.t1',
 'YF00035567.t1',
 'YF00035572.t1',
 'YF00035582.t1',
 'YF00035598.t1',
 'YF00035607.t1',
 'YF00035613.t1',
 'YF00035616.t1',
 'YF00035623.t1',
 'YF00035629.t1',
 'YF00035643.t1',
 'YF00035647.t1',
 'YF00035650.t1',
 'YF00035655.t1',
 'YF00035665.t1',
 'YF00035667.t1',
 'YF000356

In both cases, the differing genes are unique YF_IDs.

### THE INTERSECTION BETWEEN THE PEPTIDE AND CDS FILE

In [7]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_pep.fasta', 'r') as f:
    pep_records = list(SeqIO.parse(f, 'fasta'))

pep = [r.id for r in pep_records]
pep[0]
pep[0], len(pep), len(set(pep))

('YF00013476.t1', 25987, 25987)

In [8]:
len(set.intersection(set(pep), set(cds))), len(set(pep)), len(set(cds))

(25987, 25987, 25987)

The genes in the two files are identical.

# NEW 2024 MODEL (TTHERM_IDs) (MAY 15)

In [9]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_CDS.fasta', 'r') as f:
    cds_may15_records = list(SeqIO.parse(f, 'fasta'))

cds_may15 = [r.id for r in cds_may15_records]
cds_may15[0]
cds_may15[0], len(cds_may15), len(set(cds_may15))

('TTHERM_00161861', 26686, 26686)

In [10]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_mRNA.fasta', 'r') as f:
    mrna_may15_records = list(SeqIO.parse(f, 'fasta'))

mrna_may15 = [r.id for r in mrna_may15_records]
mrna_may15[0], len(mrna_may15), len(set(mrna_may15))

('TTHERM_00161861', 53372, 53372)

### THE DIFFERENCE BETWEEN THE EXON AND CDS FILES

In [11]:
len(set.difference(set(mrna_may15), set(cds_may15))), len(set(mrna_may15)), len(set(cds_may15))

(26686, 53372, 26686)

The mrna file contains twice as many genes.

In [12]:
set.difference(set(mrna_may15), set(cds_may15))

{'g19154.t1',
 'g9546.t1',
 'g29543.t1',
 'g22155.t1',
 'g28151.t1',
 'g8449.t1',
 'g11271.t1',
 'g7839.t1',
 'g16727.t1',
 'g5044.t1',
 'g12908.t1',
 'g27651.t1',
 'g26074.t1',
 'g26767.t1',
 'g23604.t1',
 'g19081.t1',
 'g14307.t1',
 'g6641.t1',
 'g21386.t1',
 'g12987.t1',
 'g23450.t1',
 'g13270.t1',
 'g8515.t1',
 'g3584.t1',
 'g7730.t1',
 'g2374.t1',
 'g6263.t1',
 'g29159.t1',
 'g134.t1',
 'g18652.t1',
 'g15995.t1',
 'g1984.t1',
 'g27279.t1',
 'g7047.t1',
 'g8681.t1',
 'g7079.t1',
 'g3333.t1',
 'g24884.t1',
 'g11069.t1',
 'g22916.t1',
 'g1342.t1',
 'g7275.t1',
 'g16966.t1',
 'g24960.t1',
 'g4868.t1',
 'g1153.t1',
 'g25524.t1',
 'g6805.t1',
 'g25312.t1',
 'g23488.t1',
 'g13238.t1',
 'g27531.t1',
 'g28014.t1',
 'g21973.t1',
 'g3377.t1',
 'g29217.t1',
 'g8618.t1',
 'g7747.t1',
 'g11836.t1',
 'g23076.t1',
 'g23631.t1',
 'g4148.t1',
 'g20563.t1',
 'g21605.t1',
 'g17105.t1',
 'g8233.t1',
 'g16786.t1',
 'g16588.t1',
 'g8013.t1',
 'g16208.t1',
 'g12557.t1',
 'g26357.t1',
 'g19739.t1',
 'g279

All of the duplicated genes are gids. I wonder if all of the sequences are duplicated.

In [13]:
mrna_may15_g_seqs = {r.id: str(r.seq) for r in mrna_may15_records if 'g' in r.id}
# mrna_may15_ttherm_seqs = {r.id: str(r.seq) for r in mrna_may15_records if 'TTHERM_' in r.id}
# mrna_may15_ttherm_seqs = {r.id: str(r.seq.complement()) for r in mrna_may15_records if 'TTHERM_' in r.id}
# mrna_may15_ttherm_seqs = {r.id: str(r.seq)[::-1] for r in mrna_may15_records if 'TTHERM_' in r.id}
mrna_may15_ttherm_seqs = {r.id: str(r.seq.reverse_complement()) for r in mrna_may15_records if 'TTHERM_' in r.id}

In [14]:
len(mrna_may15_g_seqs), list(mrna_may15_g_seqs.keys())[0], len(mrna_may15_ttherm_seqs), list(mrna_may15_ttherm_seqs.keys())[0]

(26686, 'g18410.t1', 26686, 'TTHERM_00161861')

In [15]:
sorted(list(mrna_may15_g_seqs.values()))[0]

'AAAAAAAAAAAAAAAAATTAAGCAAGGCCTACTTTATTTTTAAAAATTTTGAAAAACTTATTTTCCATTTTATTATAAACAGAGTATGAAAACAAATTAGCTTTGTCAAATTGATATTCCTAAAAATTAATATTCAATTAATGCGCCATCAAACTAACACTATAATTATAAATATCTAAAGGAAAAAATTCTACCTGGCCTTACTCCTATAAACTCTCCTGATTTATCACCTTAGATATCATTAAATTAGATAGAAGATCCAGCATTTGCAAATATTAAAATGCTAAAAATTGATTAAGAAAGTAAGAATAATTTTTCTATGAACTAATTTTCTCAATTAAGATATGGCAGCAGAGTTATACATAAAGAACAACATAAGGCTGAGTTTGATAATCTAAATAATTGTGAGGTTCTGCTAACTATTATAAAAAATTTATACAAAATATTCAACACCCTTAAAATAATAGGCAGAAAACGATGAGAAAGAGTAATAAACAGGTATTGTACATGTTTCTAATTACCATAAAAAATAAACTGAAGCCAAAAAATCAGAAGACCTTGCTTCAGAGCTATAAGGATTTGAATAAGAAGTGTATGGTACAAGCAACAAATCTACTGAATCAAAATTTAGTGTTTTACAAAAATTTAGTACAACCAATAATAATAAATTAGAATTTCAAGACAATTCTTTAGAGTTTTACAATATGTAAATTATCTTATCTTTTTTTTACACATAAAGTTTATGGAAATTAAATTAAGTTGAATTTTATTAAACTATGACAAATAATAATTAATCATAATTTACCAATAATTTAATAGATTTAGGATTTAATTAATAGCTATGCTATGATAATGTTTTATTGTGATTTAAAAAATAATAAATATTTATCTAAAAGCTGGGTTACTTTTAAATAATTTGTTGAAAATTTAATATATTTTTTTTAAAAGTTGATTTTATTTTTATTTTGGAGTGATAAAAATAAATTTGAAAAAACATTA

In [16]:
sorted(list(mrna_may15_ttherm_seqs.values()))[0]

'AAAAAAAAAAATGAATTACAATTTTTTAACTTAAAGCTAAAAAGTAGTGCCATATAAAGGATAAAATTACAATTTAAAAAAATATAAGAAACATTTGTAATGAAATAACAAGTTTTATTCATATTGGCAATACTAGAACTAATATAGATATTTATATAATTTATGAATAAGTGTATAAGCTATTTTATTAGCTATTTTATGAGCTTTAAGTGGAAGAACTTTAACAATCTTTTTGAGTTTTAGCCGCTCCTGCTTTTGATGTTGTATTGTTTGGACATTTAATACATTAAGTTTTTGAAGCACTTGTTGTTTAACTTGCATCTCCATAAAAATTAGCAGCACATTAGCATTAATTTTTATCTACAGACCAAGTCACTGCATTTGAATCATAGCATTTACAGCCAATAAAATCTTCTGAGTTAGTAGATCCAGGACCACAATTGCTACAGGCTGTATTTCCAGTACCATTTGCATATTATCCTGAGCTACACTATGTGCAGTTTACATTAGCTGAGCCATTTGAATATGTGCCTGAGGGGCAAGGAGAACACTTTCCAGATGGTAAATCTGTAAATTATCCAGCTGGGCAAGCTGTGCAACTATTTGAATTAGCTGATATTCCAATATAACCTGGATTACAAACACAGCTATCTCCATTTAATTTAGAGTTAGGAATGCATTAACAATCAGAAAAATCTGGAGTATTTGTTTTACCAGAGCCGCAAGTTTAACATGATGTATTCCCAGTATCAGAAGCGAATTAACCGCTACCACACTACTAACATTAAGAACTTCCAGTACTAGCCGCAAAATATCCAGCAGCGCAACGAGTACATGCTCCAGAGGTAGAAGAATTTGAATACTGACCTTTTTGGCATGGTGAGCATCCTTAATAGCTTGATAAACCTGATGGATTACCCTAATAGCCGTCAGGACATACGCATCTATTGTTTTTAAATACAGCACTTGTATCTGGACACTTACAGCCAGAAAAATCTG

In [17]:
len(sorted(list(mrna_may15_g_seqs.values()))[0]), len(sorted(list(mrna_may15_ttherm_seqs.values()))[0])

(2128, 2629)

In [18]:
sorted(list(mrna_may15_g_seqs.values()))[0] == sorted(list(mrna_may15_ttherm_seqs.values()))[0]

False

In [19]:
not_equal = 0

for idx, id in tqdm.tqdm(enumerate(sorted(list(mrna_may15_g_seqs.values())))):
    if id != sorted(list(mrna_may15_ttherm_seqs.values()))[idx]:
        not_equal += 1

not_equal

26686it [03:53, 114.46it/s]


26686

The sequences are not duplicated.

### THE INTERSECTION BETWEEN THE PEPTIDE AND CDS FILE

In [20]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_Protein.fasta', 'r') as f:
    pep_may15_records = list(SeqIO.parse(f, 'fasta'))

pep_may15 = [r.id for r in pep_may15_records]
pep_may15[0], len(pep_may15), len(set(pep_may15))

('TTHERM_00161861', 26686, 26686)

In [21]:
len(set.intersection(set(pep_may15), set(cds_may15))), len(set(pep_may15)), len(set(cds_may15))

(26686, 26686, 26686)

The genes in the two files are identical.